### DISCUSSION TITLE
Default featured image

### DISCUSSION THREAD

**Discussion by milesdame** [wprig/wprig#730] [Q&A] [answered]
Hello, 

The theme I am building needs to have a default featured image so that if someone creates a post without a featured image it won't mess up the design. I found several sites that discuss adding the following code block to allow for a featured image, but I'm not sure where to add it or how to incorporate it into wprig. 

```
<?php if ( has_post_thumbnail() ) {
the_post_thumbnail();
} else { ?>
<img src="<?php bloginfo('template_directory'); ?>/images/preset-featured-image.jpg" alt="<?php the_title(); ?>" />
<?php } ?>
```
I'm not very familiar with php so I've been trying to learn as I go. Right now I've attempted to insert it into `entry_thumbnail.php` right after the else statement in the if statement that checks if it's a singular post. I've obviously modified the code above to wrap around the existing the_post_thumbnail() . So far this has not worked and the image file I saved in the directory is not being displayed on my test post.

**Comment by jacklowrie** [comment]
If you inspect the html output, is the `img` tag there? It's possible the path isn't correct.

**Reply by milesdame** [reply]
No the `img` tag does show up. I've tried to make sure the bath is correct. I added `/assets` to the code snippet above and that didn't seem to affect it either.

**Reply by jacklowrie** [reply]
Ok - what does the URL look like in the markup? you could try trace that path in your folder structure to see if the file is there. If it isn't, maybe try running `npm run build`? If you added the image to `/assets/images/src` without doing so, then it may not exist yet. If you added it to `/assets/images` (as you have the path), it won't get checked into version control by default.

another thing that might help debug (or make the template a little easier to read) is to build the URL and assign it to a descriptively named variable, then add it to the html. something like this:
```php
<?php
if ( has_post_thumbnail() ) {
    the_post_thumbnail();
} else {
    $default_image_url = get_stylesheet_directory_uri() . '/images/preset-featured-image.jpg';
    ?>
    <img src="<?php esc_url( $default_image_url ); ?>" alt="<?php the_title(); ?>" />
    <?php
}
?>
```
that way, you can also debug by printing out `$default_image_url` to make sure it has the right value. something like adding `var_dump($default_image_url);` after building it.

**Comment by dthenley** [comment]
@milesdame are you inserting your new image in the folder in the theme under `assets/images/src`? if you are i would suggest changing the `img code to <img src="<?php echo get_stylesheet_directory_uri(); ?>/assets/images/preset-featured-image.jpg" alt="<?php the_title(); ?>" />`

When you `echo get_stylesheet_directory_uri()` it points to the root folder of your theme. You can place your placeholder image anywhere in the theme but I would suggest placing it in the `/assets/images/src` because when you run the wprig theme `npm run dev` command it will then auto compress that image for you and place it in the `/assets/images/` folder for you

**Reply by milesdame** [reply]
Yes I placed it in that folder. I tried to adjust the `img` code like you suggested but nothing changed. The `get_stylesheet_directory_uri()` is giving me a message that all output should be run through an escaping function. I've tried esc_html and esc_attr but I'm not sure I'm writing them correctly since I'm not experienced with php.

**Reply by milesdame** [reply]
What I'm currently attempting is this:

```
<img src="<?php echo esc_attr( get_template_directory_uri() ); ?>/assets/images/preset-featured-image.jpg" alt="<?php the_title(); ?>" />
```

**Reply by jacklowrie** [reply]
^ for this, you'd want `esc_url`. you can read more about how wordpress handles escaping output [here](https://developer.wordpress.org/themes/theme-security/data-sanitization-escaping/#escaping-securing-output). Doublecheck what that looks like when rendered, you may have a double slash - can't remember how esc_url deals with trailing slashes.

**Comment by LukaszJaro** [comment, marked as answer]
I implemented this for the archive pages, although a fallback image can also be added to the single post if you like but then you should also consider adding it for the latest news post widget/block.

You were right to insert in entry_thumbnail.php

Here's the relevant code:

```
if ( post_password_required() || ! post_type_supports( $support_slug, 'thumbnail' ) ) {
	return;
}

if ( ! has_post_thumbnail() && is_page() ) {
	return;
}

if ( is_single() ) {
	?>
	<div class="post-thumbnail">
		<?php the_post_thumbnail( 'wp-rig-featured', array( 'class' => 'skip-lazy' ) ); ?>
	</div><!-- .post-thumbnail -->
	<?php
} else {
	?>
	<div class="post-thumbnail">
		<?php
		if ( has_post_thumbnail() ) {
			the_post_thumbnail(
				'post-thumbnail',
				array(
					'alt' => the_title_attribute(
						array(
							'echo' => false,
						)
					),
				)
			);
		} else {
			echo '<img loading="lazy" width="320" height="200" src="' . esc_attr( get_template_directory_uri() ) . '/assets/images/fallback.png" class="size-post-thumbnail attachment-post-thumbnail" aria-hidden="true">';
		}
		?>
	</div><!-- .post-thumbnail -->
	<?php
}
```

I particularly had issues with this line causing the fall back image to not display:

```
if ( post_password_required() || ! post_type_supports( $support_slug, 'thumbnail' ) || ! has_post_thumbnail() ) {
	return;
}
```

**Reply by LukaszJaro** [reply]
Good catch @jacklowrie to use esc_url, I realized I used esc_attr.
